### PR TITLE
add random.multivariate_normal, fix empty_like dtype problem, fix gat…

### PR DESCRIPTION
**Pull request by Alicia1529**
1. add operator random.multivariate_normal

2. modify the test for gather_nd check bound: change the way to raise error

3. fix the infer_shape problem in custom op: empty_like, and add more dtype support

## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/ndarray/numpy/random.py [hunk 1]
```diff
@@ -245,6 +246,84 @@ def multinomial(n, pvals, size=None):
...
+        Otherwise, the output shape is ``size + np.broadcast(mean.shape, cov.shape[:-1])``
+
+    Examples
+    --------
+    >>> mean = np.array([1, 2])
+    >>> cov = np.array([[1, 0], [0, 1]])
+    >>> x = np.random.multivariate_normal(mean, cov, (3, 3))
+    >>> x.shape
+    (3, 3, 2)
+    # Performs autobroadcasting when the batch shape of
```
Comment (xidulu) [inline python/mxnet/ndarray/numpy/random.py]: I think the  `#` in this line and below are unnecessary.

File: python/mxnet/numpy_op_fallback.py [hunk 2]
```diff
@@ -158,3 +165,62 @@ def infer_shape(self, in_shape):
...
+class MultivariateNormalProp(operator.CustomOpProp):
+    """Fallback np.random.multivariate_normal operator properties."""
+
+    def __init__(self, size=None):
+        super(MultivariateNormalProp, self).__init__(need_top_grad=True)
+        self._size = ast.literal_eval(
+            size) if size is not None else None
+
+    def list_arguments(self):
+        return ['mean', 'variance']
```
Comment (xidulu) [inline python/mxnet/numpy_op_fallback.py]: Better change this to `['mean', 'cov']`

File: python/mxnet/numpy_op_fallback.py [hunk 3]
```diff
@@ -158,3 +165,62 @@ def infer_shape(self, in_shape):
...
+
+    def infer_shape(self, in_shape):
+        loc_shape = in_shape[0]
+        cov_shape = in_shape[1]
+        if len(loc_shape) < 1:
+            raise ValueError("mean must be at least 1 dimensional")
+        if len(cov_shape) < 2:
+            raise ValueError("cov must be at least 2 dimensional")
+        if cov_shape[-1] != cov_shape[-1]:
```
Comment (xidulu) [inline python/mxnet/numpy_op_fallback.py]: Do you mean `cov_shape[-1] != cov_shape[-2]:` ?

File: python/mxnet/ndarray/numpy/random.py [hunk 4]
```diff
@@ -245,6 +246,84 @@ def multinomial(n, pvals, size=None):
         return _npi.multinomial(n=n, pvals=pvals, size=size)
 
 
+def multivariate_normal(mean, cov, size=None, check_valid=None, tol=None):
+    """
+    multivariate_normal(mean, cov, size=None,  check_valid=None, tol=None)
```
Comment (xidulu) [inline python/mxnet/ndarray/numpy/random.py]: Remove the extra blank between `size=None` and `check_valid`

File: python/mxnet/numpy_op_fallback.py [hunk 5]
```diff
@@ -158,3 +165,62 @@ def infer_shape(self, in_shape):
...
+        if cov_shape[-1] != loc_shape[-1]:
+            raise ValueError("mean and cov must have same length."
+                             "The shape of mean is {} but the shape of cov is {}"
+                             .format(loc_shape[-1:], cov_shape[-2:]))
+        # handle shape mismatch here
+        out_shape = np.broadcast(np.empty(loc_shape), np.empty(cov_shape[:-1])).shape
+        if self._size is not None:
+            self._size = [self._size] if np.isscalar(
+                self._size) else list(self._size)
+            out_shape = self._size + list(out_shape)
```
Comment (xidulu) [inline python/mxnet/numpy_op_fallback.py]: You could directly concat two tuples together like `(1,2) + (3,4)`

File: python/mxnet/numpy/random.py [hunk 6]
```diff
@@ -242,6 +242,82 @@ def multinomial(n, pvals, size=None, **kwargs):
...
+    array([[-1.6115597 , -8.726251  ],
+           [ 2.2425299 ,  2.8104177 ],
+           [ 0.36229908, -8.386591  ]])
+
+    The following is probably true, given that 0.6 is roughly twice the
+    standard deviation:
+    >>> list((x[0,0,:] - mean) < 0.6)
+    [True, True] # random
+    """
+    return _mx_nd_np.random.multivariate_normal(mean, cov, size=size,\
```
Comment (haojin2) [inline python/mxnet/numpy/random.py]: no need for backslash here, and fix alignment

File: python/mxnet/numpy_op_fallback.py [hunk 7]
```diff
@@ -158,3 +165,62 @@ def infer_shape(self, in_shape):
 
     def create_operator(self, ctx, in_shapes, in_dtypes):
         return Unravel_index(self._shape)
+
```
Comment (haojin2) [inline python/mxnet/numpy_op_fallback.py]: one more blank line.

Comment (haojin2) [inline python/mxnet/numpy_op_fallback.py]: one more blank line.

File: python/mxnet/numpy/random.py (hunk 6 above)
Comment (haojin2) [inline python/mxnet/numpy/random.py]: no need for backslash here, and fix alignment
